### ISSUE TITLE
[question] Get dependencies of a package in custom generator

### ISSUE THREAD

**Issue by Sil3ntStorm** [conan-io/conan#11637]
Previously I was able to get the dependencies of conan packages in a custom generator using something like

```
        for name, info in self.deps_build_info.dependencies:
            if not name in self.conanfile.build_requires:
                if info.components:
                    for _, comp in info.components.items():
                        self.processSingleDependency(comp)
                else:
                    result = self.processSingleDependency(info)
```

which would then attempt to read the things that said dependency depends on, among other needed information.
This still works for things like abseil, grpc (which use components) and seemingly (didn't take a close look yet) a few other required packages.
But it does no longer seem to work for things like json-schema-validator (which depends on nlohmann_json) or sqlitecpp (which depends on sqlite3).

```
    def processSingleDependency(self, info):
        if info.lib_paths:
            self.conanfile.output.info(f'  Lib Directories: {info.lib_paths}')
        if info.requires:
            self.conanfile.output.info(f'  Required packages: {info.requires}')
        if info.include_paths:
            self.conanfile.output.info(f'  Include Directories: {info.include_paths}')
```

It does appear that requires is always empty, despite it actually requiring the dependency to be present (especially in the case of sqlitecpp).

Conan itself seems perfectly aware of the package requiring another package as evidenced by `conan info` and the requirements are in the dependency tree.

Has anything changed in recent history?
How are custom generators supposed to gain the information as to the requirements of libraries currently?

Using conan 1.50

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
Please, use ``self.dependencies`` to access the dependencies information for generators: https://docs.conan.io/en/latest/reference/conanfile/dependencies.html

For writing 2.0 generators:
- Do not extend ``Generator`` class
- Use ``python_requires`` to put and share the code
- Instantiate it explicitly in the consumers ``generate()`` method
- use ``self.dependencies`` to access information

**Comment by Sil3ntStorm** [comment]
I just cannot seem to win here and am growing increasingly frustrated with this...
What I was doing worked fine for all the recipes I ever tried for like the past year, and it appears to still work fine, except it doesn't for the sqlitecpp recipe.

The [documentation](https://docs.conan.io/en/latest/reference/conanfile/dependencies.html#dependencies-cpp-info-interface) mentions:

```
cpp_info.includedirs
cpp_info.libdirs
```

which doing so for me will simply result in an error about NewCppInfo not having such a thing...
It appears the only available things are
```
    'aggregated_components',
    'clear_none',
    'component_names',
    'components',
    'get_sorted_components',
    'has_components',
    'merge',
    'required_components',
    'set_relative_base_folder'
```    
None of which I seem to be able to access properly, as there is no apparent way to use them and attempting to discover what they offer just fails in various ways.
Not to mention that has_components is false, yet all there is is components stuff.

**All I want is being able to tell that sqlitecpp requires sqlite3 FFS, but apparently that seems impossible.**

Of course there is the requirement of it being universal and must work for all CCI recipes when using conan 1.50.

**Comment by memsharded** [comment]
Hi @Sil3ntStorm 

I don't know where the error is coming from. You say that ``NewCppInfo`` does not have ``includedirs``, but it does. The only thing it is evaluated dynamically, via ``__getattr__`` implementation. We use ``cpp_info.includedirs`` extensively in all of our new generators without any issue. I have checked it and it is also present in some tests in the test suite.

The main issue is that we are missing information to reproduce. It is not enough to say that it is failing in different ways, the best way is to provide some "minimal reproducible case", typically in a git repo, or just here, but the guidelines are:
- It should be as minimal as possible, simplify all that is not strictly necessary to reproduce
- It should be as self-contained as possible. It is possible to generate "hello" dependencies to try things with ``conan new hello/0.1 -m=cmake_lib && conan create .``
- It should be complete, including all necessary code to reproduce, and all the exact commands to reproduce.


> All I want is being able to tell that sqlitecpp requires sqlite3 FFS, but apparently that seems impossible.

Please avoid these kind of expressions in our repos. I am sincerely sorry that you are frustrated, and we are always trying to help, but this doesn't help. I don't think that this would be impossible, because the ``self.dependencies`` model is far more complete and powerful than the previous one. And if there is something that it is not possible today but make sense, we will work as always to add it. But we need to understand well the issues and be able to reproduce, otherwise it is impossible to move forward.

**Comment by Sil3ntStorm** [comment]
You're correct, I tend to not control very much what I say / write, even less so when frustrated with things that I cannot seem to get to work. Sorry about that.

I put together https://github.com/Sil3ntStorm/conan-issues/tree/custom-generator-1 
Not sure if that is any good, but it does illustrate the issue. Simply run `python3 repro.py bad` and examine the `conanbuildinfo.dummy` file.
There does not seem to be any information that I can see which would suggest that sqlitecpp needs to link against sqlite3 from the information printed. As such the generator is unable to tell (as far as I am aware) that you need to link to sqlite3 in order to link successfully.

There is also `python3 repro.py good` that simply requires grpc. It's quite large, however does depend on a few third party libraries and produces output that allows the generator to tell, what needs what in order to successfully link.

However I am starting to think that this might be an issue with the sqlitecpp recipe? If so I am perfectly happy to close this issue and create a new one in CCI.

**Comment by memsharded** [comment]
Hi @Sil3ntStorm 

Sorry that this issue went staled and we didn't follow up, specially when you provided a repo to reproduce.
Now it seems this ticket would be quite outdated, it was referring to legacy Conan 1.X, deprecated more than 2 years ago by Conan 2, and also about legacy generators even within Conan 1.X.

So I am closing this ticket as outdated, please don't hesitate to create new tickets if necessary, nowadays we are able to keep up a bit better with incoming support tickets (still at peak times some issue might fall through the cracks, don't hesitate to ping on any of them to re-surface them, that also help us). Thanks very much for your feedback.
